### PR TITLE
RM-8652 Remove AppDomainRunner usages and make it .NET Framework only

### PR DESCRIPTION
**Pull request by a-ctor**
https://re-motion.atlassian.net/browse/RM-8652

### CODE REVIEW

File: Remotion/Core/UnitTests/Reflection/ContextAwareTypeUtilityTest.cs [hunk 1]
```diff
@@ -122,23 +122,16 @@ public void GetTypeResolutionService_Cached ()
...
-#if NETFRAMEWORK
     [Test]
     public void GetTypeDiscoveryService_WithCustomImplementation_DoesNotThrowOnInitialization ()
     {
-      var relativePath = @"Reflection\TestDomain\ContextAwareTypeUtilityTest\app.config";
-      var fullPath = Path.Combine(TestContext.CurrentContext.TestDirectory, relativePath);
-      Assert.That(File.Exists(fullPath));
+      var typeDiscoveryConfiguration = new TypeDiscoveryConfiguration();
+      typeDiscoveryConfiguration.Mode = TypeDiscoveryMode.CustomTypeDiscoveryService;
+      typeDiscoveryConfiguration.CustomTypeDiscoveryService.Type = typeof(FakeTypeDiscoveryService);
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Reflection/ContextAwareTypeUtilityTest.cs:130]: This test ensures that RM-8047 works correctly (see also RM-8064). We need to improve the naming. It is mandatory that we use the config-files, not intialization by assignment because the type name must be deserialized. The entire setup runs in an isolated AppDomain to ensure the statics are unintialized when things start.

**Review by MichaelKetting — COMMENTED** [review]
partial review

File: Remotion/Core/UnitTests/Reflection/ContextAwareTypeUtilityTest.cs (hunk 1 above)
Comment (MichaelKetting) [reply Remotion/Core/UnitTests/Reflection/ContextAwareTypeUtilityTest.cs:130]: Note: if the loadcontext works to re-initialize the statics, loadcontext is still a valid approach

File: Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyFinding/AssemblyFinderTest.cs [hunk 2]
```diff
@@ -214,11 +211,10 @@ private bool IsAssemblyReferencedBy (Assembly referenced, Assembly origin)
           .Any();
     }
 
-    private static Assembly CompileReferencingAssembly (AssemblyCompilerBuildOutputManager outputManager, Assembly remotionAssembly)
+    private static string CompileReferencingAssembly (AssemblyCompilerBuildOutputManager outputManager, Assembly remotionAssembly)
     {
       var referencingAssemblyRelativePath = outputManager.CompileInSeparateAppDomain("RemotionCoreReferencingAssembly.dll", remotionAssembly.Location);
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyFinding/AssemblyFinderTest.cs:216]: When switching the compiler-logic, we need to check if CompileInSeparateAppDomain is still relevant. The basic idea was to prevent the compiled Assembly from getting loaded by accident. It's possible that the CSharpCodeProvider did load the assembly after compilation. If that's not even the case right now, we could also drop this api immediately since here is the only usage.

File: Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs [hunk 3]
```diff
@@ -82,10 +82,9 @@ public void AddIgnoredAssembly ()
...
     }
 
+#if NETFRAMEWORK
     [Test]
-#if !NETFRAMEWORK
     [Ignore("TODO RM-7799: Create out-of-process test infrastructure to replace tests done with app domains; RM-7808: Integrate the RoslynCodeDomProvider and renable the AssemblyCompiler tests")]
-#endif
     public void ApplicationAssemblyInclusion_DependsOnAttribute ()
     {
       string compiledAssemblyPath = Path.Combine(AppContext.BaseDirectory, "NonApplicationMarkedAssembly.dll");
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs:90]: I'm not sure about this.

1.  you disabled the test also for .net framework, leaving the logic untested until the next task is implemented in a few weeks. That's a bit risky, although right now, there's every expectation that we will implement RM-7808 before the release.

2. Could we just switch this to assemblyloadcontext instead of appdomainrunner? Can we instantiate/execute a type in a new assemblyloadcontext in .net?

File: Remotion/Development/Core/Development.Core.csproj [hunk 4]
```diff
@@ -12,9 +12,7 @@
     <RootNamespace>$(AssemblyName)</RootNamespace>
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\**\*"
-             Link="SharedSource\UnitTesting\%(RecursiveDir)%(Filename)%(Extension)"
-             Exclude="**\PEVerifyVersion.cs;**\NUnit\**" />
+    <Compile Include="..\..\..\SharedSource\Development\UnitTesting\**\*" Link="SharedSource\UnitTesting\%(RecursiveDir)%(Filename)%(Extension)" Exclude="**\PEVerifyVersion.cs;**\NUnit\**" />
```
Comment (MichaelKetting) [inline Remotion/Development/Core/Development.Core.csproj:15]: This is just newline formatting? is this change needed/helpful?

**Review by MichaelKetting — COMMENTED** [review]
partial review

File: Remotion/Data/DomainObjects.UnitTests/ReflectionUtilityTest.cs [hunk 5]
```diff
@@ -83,12 +80,21 @@ public void GetAssemblyPath_WithHashInDirectoryName ()
...
             directoryPath,
             newAssemblyPath);
+#else
+        using var assemblyContext = new UnloadableAssemblyContext();
+        assemblyContext.RunWithAssemblyLoadContext(
+            context =>
+            {
+              var assembly = context.LoadFromAssemblyPath(newAssemblyPath);
+              Assert.That(Path.GetDirectoryName(assembly.Location), Is.EqualTo(directoryPath));
+              Assert.That(ReflectionUtility.GetAssemblyDirectory(assembly), Is.EqualTo(Path.TrimEndingDirectorySeparator(AppContext.BaseDirectory)));
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/ReflectionUtilityTest.cs:90]: It feels like we should de-duplicate the logic in the test. It's just the knowledge that .NET Framework will die in less than a year that I'm ambivalent about changing it now.

File: Remotion/Data/DomainObjects.UnitTests/Queries/Configuration/QueryFileElementTest.cs [hunk 6]
```diff
@@ -49,22 +50,27 @@ public void GetRootedPath_WithUnrootedPath_ReturnsPathRelativeToAppBase ()
...
         string path = @"foo\bar.txt";
         string fullPath = Path.Combine(AppContext.BaseDirectory, @"foo\bar.txt");
         Assert.That(QueryFileElement.GetRootedPath(path), Is.EqualTo(fullPath));
       }).Run();
+#else
+      using (new ChangedApplicationBaseDirectorySection(@"c:\"))
+      {
+        string path = @"foo\bar.txt";
+        string fullPath = Path.Combine(AppContext.BaseDirectory, @"foo\bar.txt");
+        Assert.That(QueryFileElement.GetRootedPath(path), Is.EqualTo(fullPath));
```
Comment (MichaelKetting) [inline Remotion/Data/DomainObjects.UnitTests/Queries/Configuration/QueryFileElementTest.cs:71]: See ReflectionUtilityTest on comment about deduplication

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs (hunk 3 above)
Comment (a-ctor) [reply Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs:90]: Something went wrong here. The problem is that I cannot test that I correctly converted the test to `AssemblyLoadContext` as the `AssemblyCompiler` is not working yet. Since the `AppDomainRunner` is now NETFRAMWORK-only parts of this test need to be conditional. I suggest throwing a `NotImplementedException` and re-adding the Ignore test for .NET Core.

File: Remotion/Development/Core/UnitTesting/UnloadableAssemblyContext.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    private AssemblyLoadContext? _assemblyLoadContext;
+
+    public UnloadableAssemblyContext ()
+    {
+      _assemblyLoadContext = new AssemblyLoadContext(null, true);
+      _weakAssemblyLoadContextReference = new WeakReference(_assemblyLoadContext);
+    }
+
+    internal AssemblyLoadContext Context => _assemblyLoadContext ?? throw new InvalidOperationException("Assembly load context has been unloaded.");
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Development/Core/UnitTesting/UnloadableAssemblyContext.cs:41]: It is used inside the class but right now could be made `private`.

File: Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs (hunk 3 above)
Comment (MichaelKetting) [reply Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs:90]: Or option two removes the requirement for AppDomainRunner and solves all other issues, too? Question is, if it also solves this for .NET Framework?

File: Remotion/Development/Core/UnitTesting/UnloadableAssemblyContext.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+using System.Reflection;
+using System.Runtime.Loader;
+using System.Threading;
+
+namespace Remotion.Development.UnitTesting
+{
+  /// <summary>
+  /// Wraps an unloadable <see cref="AssemblyLoadContext"/> with the dispose pattern.
+  /// </summary>
+  public class UnloadableAssemblyContext : IDisposable
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Development/Core/UnitTesting/UnloadableAssemblyContext.cs:29]: `AssemblyLoadContext` does not implement `IDisposable`. It can be marked as "unloadable" which can then be triggered using `.Unload()`. The thing is that it just removes the "keep alive" reference to the `AssemblyLoadContext` but does not wait for the actual unload. This class is supposed to provide that with the using pattern.

File: Remotion/Data/DomainObjects.UnitTests/ReflectionUtilityTest.cs (hunk 5 above)
Comment (a-ctor) [reply Remotion/Data/DomainObjects.UnitTests/ReflectionUtilityTest.cs:90]: In this case there are differences between .NET Framework and .NET Core so I would not try to deduplicate the logic here. I think that the de-duplication would add more complexity than the conditional. We don't have many places where we use conditional code.

File: Remotion/Development/Core/UnitTesting/UnloadableAssemblyContext.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Development/Core/UnitTesting/UnloadableAssemblyContext.cs:29]: so, better summary it is, then. "AssemlyLoadContext that waits until the unload operation has completed" or something similiar/better.

File: Remotion/Data/DomainObjects.UnitTests/ReflectionUtilityTest.cs (hunk 5 above)
Comment (MichaelKetting) [reply Remotion/Data/DomainObjects.UnitTests/ReflectionUtilityTest.cs:90]: Ah, the assembly-load is also different. hmm... okay.

File: Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs (hunk 3 above)
Comment (a-ctor) [reply Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs:90]: `AssemblyLoadContext` only exists in .NET Core so I can only run the test in .NET Core where the `AssemblyCompiler` is broken. Once the `AssemblyCompiler` is repaired the .NET Core test based of `AssemblyLoadContext` can be implemented. So for now I would only add a `NotImplementedException` in the .NET Core case to remind the person that enables the test again that there is something left to do.

File: Remotion/Mixins/MixerTools.UnitTests/Mixer_IntegrationTest.cs [hunk 9]
```diff
@@ -51,41 +51,37 @@ public void TearDown ()
     [Ignore("TODO RM-7799: Create out-of-process test infrastructure to replace tests done with app domains")]
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Mixins/MixerTools.UnitTests/Mixer_IntegrationTest.cs:51]: Might be disabled for a reason:

```
Remotion.Reflection.TypeDiscovery.AssemblyLoading.AssemblyLoaderException : The assembly 'Microsoft.VisualStudio.TestPlatform.ObjectModel, Version=11.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' (loaded in the context of 'NUnit3.TestAdapter, Version=4.0.0.0, Culture=neutral, PublicKeyToken=4cb40d35494691ac') triggered a FileNotFoundException - maybe the assembly does not exist or a referenced assembly is missing?
FileNotFoundException message: Could not load file or assembly 'Microsoft.VisualStudio.TestPlatform.ObjectModel, Version=15.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
```

Since the test won't run I cannot test if the assemblies are loaded. This should probably be a separate issue to resolve this error and remove the ignore. Since the mixer project is .NET Framework right now I would just revert the changes and once we convert the project to .NET we can rewrite the test

File: Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/Development.UnitTesting.IsolatedCodeRunner.UnitTests.csproj [hunk 10]
```diff
@@ -0,0 +1,23 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <ProjectType>Test</ProjectType>
+    <TargetFrameworks></TargetFrameworks>
+    <TargetFramework>net6.0-windows</TargetFramework>
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/Development.UnitTesting.IsolatedCodeRunner.UnitTests.csproj:6]: I think we should consider adding a flag for this. right now, it's only in WebApplication.props and now here. If we get more projects that are .net only, having a single declaration in remotion might be helpful. 

Also, the target frameworks override is declared before Shared.build.props, which means that they will get overridden again. This also gets solved by introducing a flag. `UseOnlyDotNetAsTargetFramework` or something of the sort. Not happy yet with the name.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
+﻿using System;
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs:1]: license header
-> also a good analyzer rule.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    private static bool _shouldStayFalse;
+
+    [Test]
+    public void Run_ExecutesCodeInNewProcess ()
+    {
+      var appProcessRunner = new IsolatedCodeRunner(TestAction);
+      appProcessRunner.Run();
+
+      Assert.That(appProcessRunner.Output.Trim(), Is.EqualTo("ran"));
+      Assert.That(_shouldStayFalse, Is.EqualTo(false));
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs:20]:
```suggestion
      Assert.That(_shouldStayFalse, Is.False);
```
convention

File: Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+{
+  [TestFixture]
+  public class IsolatedCodeRunnerTests
+  {
+    private static bool _shouldStayFalse;
+
+    [Test]
+    public void Run_ExecutesCodeInNewProcess ()
+    {
+      var appProcessRunner = new IsolatedCodeRunner(TestAction);
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs:16]: variable naming

File: Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+    [Test]
+    public void Run_ArgumentsArePassedToNewProcess ()
+    {
+      var appProcessRunner = new IsolatedCodeRunner(TestAction);
+      appProcessRunner.Run("a", "b", "c");
+
+      static void TestAction (string[] args)
+      {
+        Assert.That(args, Is.EquivalentTo(new[] { "a", "b", "c" }));
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs:53]:
```suggestion
        Assert.That(args, Is.EqualTo(new[] { "a", "b", "c" }));
```
The order of items is important.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/Development.UnitTesting.IsolatedCodeRunner.csproj [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <ProjectType>Executable</ProjectType>
+    <TargetFrameworks></TargetFrameworks>
+    <TargetFramework>net6.0-windows</TargetFramework>
+  </PropertyGroup>
+  <Import Project="..\..\..\Build\Shared.build.props" />
+  <PropertyGroup>
+    <AssemblyTitle>re-motion Development Isolated Code Runner Library</AssemblyTitle>
+    <Description>Unit test tool for running code isloated in an external process.</Description>
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/Development.UnitTesting.IsolatedCodeRunner.csproj:11]:
```suggestion
    <Description>Unit test tool for running code in an external process to allow complete isolation of the tested system.</Description>
```

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/Development.UnitTesting.IsolatedCodeRunner.csproj [hunk 16]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <ProjectType>Executable</ProjectType>
+    <TargetFrameworks></TargetFrameworks>
+    <TargetFramework>net6.0-windows</TargetFramework>
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/Development.UnitTesting.IsolatedCodeRunner.csproj:6]: Same as in Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/Development.UnitTesting.IsolatedCodeRunner.UnitTests.csproj

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeException.cs [hunk 17]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using Remotion.Utilities;
+
+namespace Remotion.Development.UnitTesting.IsolatedCodeRunner
+{
+  public class IsolatedCodeException : Exception
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeException.cs:6]: xml doc

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Remotion.Utilities;
+
+namespace Remotion.Development.UnitTesting.IsolatedCodeRunner
+{
+  public class IsolatedCodeRunner
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:14]: xmldoc

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeException.cs [hunk 19]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeException.cs:1]: lic

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 20]
```diff
@@ -0,0 +1,167 @@
+﻿using System;
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:1]: lic

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+    public IsolatedCodeRunner (Action<string[]> testAction)
+        : this(testAction, TimeSpan.FromSeconds(15))
+    {
+    }
+
+    public IsolatedCodeRunner (Action<string[]> testAction, TimeSpan timeout)
+    {
+      ArgumentUtility.CheckNotNull(nameof(testAction), testAction);
+
+      _timeout = timeout;
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:46]: move after arg checks.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+
+      var testActionMethod = testAction.Method;
+      if (!testActionMethod.IsStatic)
+        throw new ArgumentException("The specified test method must be static.");
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:50]: this one needs a unit test

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 23]
```diff
@@ -0,0 +1,167 @@
...
+      _typeName = testActionMethod.DeclaringType!.FullName!;
+      _methodName = testActionMethod.Name;
+    }
+
+    public string Output
+    {
+      get
+      {
+        lock (_output)
+          return _output.ToString();
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:63]: braces

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs (hunk 18 above)
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:14]: we should seal this type (and the others in this assembly, too). I know, in the past I rarely sealed and I mostly wet to sealing here because of the Run method's assembly lookup. still feels like a good idea, though.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+
+      var startInfo = new ProcessStartInfo
+                      {
+                          WindowStyle = ProcessWindowStyle.Hidden,
+                          CreateNoWindow = false,
+                          RedirectStandardOutput = true
+                      };
+
+      startInfo.FileName = appProcessRunnerExe;
+      startInfo.Arguments = string.Join(" ", args.Select(e => $@"""{e.Replace("\"", "\"\"")}"""));
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:83]: we should add tests for this.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+
+      if (process.WaitForExit((int)_timeout.TotalMilliseconds))
+      {
+        // Need to call .WaitForExit() again to make sure that standard out is flushed
+        process.WaitForExit();
+      }
+      else
+      {
+        process.Kill();
+        throw new InvalidOperationException("Isolated code runner execution failed to stop within the specified timeout.");
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:110]: we can test this. Test should be marked as `LongRunning`

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 26]
```diff
@@ -0,0 +1,167 @@
...
+      }
+
+      if (_exception != null)
+        throw _exception;
+
+      if (process.ExitCode != 0)
+        throw new InvalidOperationException($"Isolated code runner execution failed with exit code {process.ExitCode}.");
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:117]: can we test this?

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 27]
```diff
@@ -0,0 +1,167 @@
...
+    }
+
+    private void DataReceived (object sender, DataReceivedEventArgs args)
+    {
+      var line = args.Data;
+      if (line == null)
+        return;
+
+      Console.WriteLine(line);
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:126]: what is this Console.Writeline doing? Just for diagnostic?

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 28]
```diff
@@ -0,0 +1,167 @@
...
+        {
+          try
+          {
+            remoteException = (Exception?)Activator.CreateInstance(Type.GetType(exceptionTypeName)!, exceptionMessage, null);
+          }
+          catch
+          {
+            // Ignore and use a IsolatedCodeException instead
+          }
+        }
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:153]: I understand what you're doing, it just feels...like we could do this with more typesafety by getting the ctor and failing/fallback to next signatur.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+
+        if (remoteException != null && !string.IsNullOrEmpty(exceptionStackTrace))
+          remoteException = ExceptionDispatchInfo.SetRemoteStackTrace(remoteException, exceptionStackTrace);
+
+        _exception = remoteException ?? new IsolatedCodeException(exceptionTypeName, exceptionMessage, exceptionStackTrace);
+      }
+
+      lock (_output)
+        _output.AppendLine(line);
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:162]: braces

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs [hunk 30]
```diff
@@ -0,0 +1,167 @@
...
+      }
+
+      if (_exception != null)
+        throw _exception;
+
+      if (process.ExitCode != 0)
+        throw new InvalidOperationException($"Isolated code runner execution failed with exit code {process.ExitCode}.");
+    }
+
+    private void DataReceived (object sender, DataReceivedEventArgs args)
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:120]: tehre is a lot of stuff in there. if we extract this into a utility class (ExceptionSerializer), we can easiliy test it. or you test it integrated via isolated. it should have 100% coverage, tough.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/Program.cs [hunk 31]
```diff
@@ -0,0 +1,93 @@
...
+
+    private static void ReportException (Exception ex)
+    {
+      // Serialization format (keep in sync with IsolatedCodeRunner.cs):
+      // ##exception:<base64 exception type name>;<base 64 exception message>;<base 64 exception stack trace>
+      var exceptionTypeName = ex.GetType().FullName;
+      var exceptionMessage = ex.Message;
+      var exceptionStackTrace = ex.StackTrace;
+
+      Console.WriteLine($"{IsolatedCodeRunner.ExceptionMarker}{EncodeBase64(exceptionTypeName)};{EncodeBase64(exceptionMessage)};{EncodeBase64(exceptionStackTrace)}");
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/Program.cs:67]: SHouldn't this formatting be a helper on the exception serialization utility?

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/Program.cs [hunk 32]
```diff
@@ -0,0 +1,93 @@
...
+      var assembly = Assembly.LoadFile(assemblyPath);
+      var type = assembly.GetType(typeName);
+      if (type == null)
+        throw new InvalidOperationException($"Could not find type '{typeName}' in assembly '{assembly.GetName()}'.");
+
+      var method = type.GetMethod(methodName, BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic);
+      if (method == null)
+        throw new InvalidOperationException($"Could not locate method '{typeName}.{methodName}' in '{assembly.GetName()}'");
+
+      return (Action<string[]>)Delegate.CreateDelegate(typeof(Action<string[]>), null, method, true)!;
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/Program.cs:90]: why not `method.CreateDelegate<Action<string[]>>`

File: Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs [hunk 33]
```diff
@@ -91,6 +91,7 @@ public void ApplicationAssemblyInclusion_DependsOnAttribute ()
...
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs:91]: The first ignore item can be removed now.

File: Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs [hunk 34]
```diff
@@ -108,13 +109,15 @@ public void ApplicationAssemblyInclusion_DependsOnAttribute ()
               assemblyCompiler.Compile();
               Assert.That(filter.ShouldIncludeAssembly(assemblyCompiler.CompiledAssembly), Is.False);
             }, compiledAssemblyPath);
+#else
+        throw new NotImplementedException();
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Reflection/TypeDiscovery/AssemblyLoading/ApplicationAssemblyLoaderFilterTest.cs:113]:
```suggestion
        throw new NotImplementedException("RM-7808: Integrate the RoslynCodeDomProvider and renable the AssemblyCompiler tests. Implement test using IsolatedCodeRunner.");
```

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Development/UnitTests/Core/UnitTesting/ChangedAppBaseDirectorySectionTest.cs [hunk 35]
```diff
@@ -0,0 +1,49 @@
...
+      }
+    }
+
+    [Test]
+    public void RestoresOldBaseDirectory ()
+    {
+      var currentBaseDirectory = AppContext.BaseDirectory;
+      using (new ChangedApplicationBaseDirectorySection("asd"))
+      {
+        // noop
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Development/UnitTests/Core/UnitTesting/ChangedAppBaseDirectorySectionTest.cs:42]: N-operation? :p

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Development/UnitTests/Core/UnitTesting/ChangedAppBaseDirectorySectionTest.cs:42]: Assembler :p
https://en.wikipedia.org/wiki/NOP_(code)

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs (hunk 26 above)
Comment (a-ctor) [reply Remotion/Development/UnitTesting.IsolatedCodeRunner/IsolatedCodeRunner.cs:117]: Not really. Since the runner will catch any exception and report it this code is only hit if something really went wrong and there is a problem reporting the exception.

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/Program.cs [hunk 36]
```diff
@@ -0,0 +1,102 @@
...
+
+    private static string? GetSettingFromEnvironment (string name)
+    {
+      return Environment.GetEnvironmentVariable(name);
+    }
+
+    private static void ReportException (Exception exception)
+    {
+      // Serialization format (keep in sync with IsolatedCodeRunner.cs):
+      // ##exception:<base64 exception type name>;<base 64 exception message>;<base 64 exception stack trace>
```
Comment (MichaelKetting) [inline Remotion/Development/UnitTesting.IsolatedCodeRunner/Program.cs:77]: that comment is now out of date, isn't it?

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

File: Remotion/Development/UnitTesting.IsolatedCodeRunner/ExceptionSerializer.cs [hunk 37]
```diff
@@ -0,0 +1,91 @@
...
+            remoteException = (Exception)messageAndInnerExceptionConstructor.Invoke(new object?[] { exceptionMessage, null });
+        }
+      }
+      catch
+      {
+        // Could not recreate the exception -> will return generic IsolatedCodeException instead
+      }
+
+      if (remoteException != null && !string.IsNullOrEmpty(exceptionStackTrace))
+        remoteException = ExceptionDispatchInfo.SetRemoteStackTrace(remoteException, exceptionStackTrace);
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Development/UnitTesting.IsolatedCodeRunner/ExceptionSerializer.cs:82]: Thats what you get for trusting a .NET API to do a sensible thing

File: Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs [hunk 38]
```diff
@@ -0,0 +1,136 @@
...
+          () => isolatedCodeRunner.Run(),
+          Throws.InvalidOperationException
+              .With.Message.EqualTo("Isolated code runner execution failed to stop within the specified timeout."));
+      sw.Stop();
+
+      Assert.That(sw.Elapsed.Seconds, Is.LessThanOrEqualTo(3));
+
+      static void TestAction (string[] args)
+      {
+        Thread.Sleep(15_000);
```
[the remark this answers is not in the document]
Comment (a-ctor) [reply Remotion/Development/UnitTesting.IsolatedCodeRunner.UnitTests/IsolatedCodeRunnerTests.cs]: I shortened the timeout to 5sec. The actual timeout is (almost) unimportant as the process is killed if it takes longer than the timeout so there should be no leftover processes.

**Review by MichaelKetting — APPROVED** [review]
